File: 백준/Silver/1926. 그림/그림.java
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.StringTokenizer;
import java.io.BufferedReader;
import java.io.IOException;

public class Main {

	static int n, m;
	static int[][] board;
	static boolean[][] visited;
	static int[] dx = {1, 0, -1, 0}, dy = {0, 1, 0, -1};
	
	private static int[] bfs() {
		int[] result = new int[2];
		Queue<int[]> q = new ArrayDeque<>();
		visited = new boolean[n][m];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (!visited[i][j] && board[i][j] == 1) {
					result[0]++;
					visited[i][j] = true;
					q.offer(new int[] {i, j});
					int width = 1;
					while (!q.isEmpty()) {
						int[] pos = q.poll();
						for (int d = 0; d < 4; d++) {
							int nx = pos[0] + dx[d];
							int ny = pos[1] + dy[d];
							if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
							if (visited[nx][ny] || board[nx][ny] == 0) continue;
							width++;
							visited[nx][ny] = true;
							q.offer(new int[] {nx, ny});
						}
					}
					result[1] = Math.max(result[1], width);
				}
			}
		}
		return result;
	}
	
	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = null;
		String[] temp = br.readLine().split(" ");
		n = Integer.parseInt(temp[0]);
		m = Integer.parseInt(temp[1]);
		board = new int[n][m];
		visited = new boolean[n][m];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < m; j++) {
				board[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		int[] result = bfs();
		System.out.println(result[0]);
		System.out.println(result[1]);
	}

}

File: 백준/Silver/1926. 그림/그림.py
from sys import stdin
from collections import deque
input = stdin.readline

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]


def out_of_bounds(x, y):
    return 0 <= x < n and 0 <= y < m


def bfs(i, j):
    area = 1
    visited[i][j] = 1
    q.append([i, j])
    while q:
        x, y = q.popleft()
        for d in range(4):
            nx = x + dx[d]
            ny = y + dy[d]
            if not out_of_bounds(nx, ny) or visited[nx][ny] or board[nx][ny] == '0':
                continue
            area += 1
            visited[nx][ny] = 1
            q.append([nx, ny])
    return area


def solve():
    global n, m, board, q, visited

    n, m = map(int, input().split())
    board = [list(input().split()) for _ in range(n)]

    q = deque()
    visited = [[0] * m for _ in range(n)]

    cnt, area = 0, 0
    for i in range(n):
        for j in range(m):
            if not visited[i][j] and board[i][j] == '1':
                cnt += 1
                area = max(area, bfs(i, j))
    print(cnt)
    print(area)


if __name__ == "__main__":
    solve()
